File: nodejs-typescript/src/domain/TransferResult.ts
import type { TransferInstruction } from './TransferInstruction.js';

export type TransferFailureReason =
  | 'source_not_found'
  | 'destination_not_found'
  | 'insufficient_funds';

export type TransferSkipReason = 'already_processed';

const REASON_MESSAGES: Record<TransferFailureReason | TransferSkipReason, string> = {
  source_not_found: 'Source account not found',
  destination_not_found: 'Destination account not found',
  insufficient_funds: 'Insufficient funds',
  already_processed: 'Transfer already processed',
};

export type TransferResult =
  | {
      readonly status: 'success';
      readonly instruction: TransferInstruction;
    }
  | {
      readonly status: 'failure';
      readonly instruction: TransferInstruction;
      readonly reason: TransferFailureReason;
      readonly message: string;
    }
  | {
      readonly status: 'skipped';
      readonly instruction: TransferInstruction;
      readonly reason: TransferSkipReason;
      readonly message: string;
    };

export function transferSuccess(instruction: TransferInstruction): TransferResult {
  return { status: 'success', instruction };
}

export function transferFailure(
  instruction: TransferInstruction,
  reason: TransferFailureReason,
): TransferResult {
  return {
    status: 'failure',
    instruction,
    reason,
    message: REASON_MESSAGES[reason],
  };
}

export function transferSkipped(
  instruction: TransferInstruction,
  reason: TransferSkipReason = 'already_processed',
): TransferResult {
  return {
    status: 'skipped',
    instruction,
    reason,
    message: REASON_MESSAGES[reason],
  };
}

export function isTransferSuccess(
  result: TransferResult,
): result is Extract<TransferResult, { status: 'success' }> {
  return result.status === 'success';
}

export function isTransferFailure(
  result: TransferResult,
): result is Extract<TransferResult, { status: 'failure' }> {
  return result.status === 'failure';
}

export function isTransferSkipped(
  result: TransferResult,
): result is Extract<TransferResult, { status: 'skipped' }> {
  return result.status === 'skipped';
}

export function transferReasonCode(result: TransferResult): string | undefined {
  if (isTransferFailure(result)) return result.reason;
  if (isTransferSkipped(result)) return result.reason;
  return undefined;
}

File: ruby/lib/mable_bank/domain/transfer_result.rb
# frozen_string_literal: true

module MableBank
  module Domain
    class TransferResult
      REASONS = {
        TransferFailureReason::SOURCE_NOT_FOUND => 'Source account not found',
        TransferFailureReason::DESTINATION_NOT_FOUND => 'Destination account not found',
        TransferFailureReason::INSUFFICIENT_FUNDS => 'Insufficient funds',
        already_processed: 'Transfer already processed'
      }.freeze

      attr_reader :instruction, :reason, :reason_code

      def self.success(instruction)
        new(instruction: instruction, status: :succeeded)
      end

      def self.failure(instruction, reason_code:)
        new(
          instruction: instruction,
          status: :failed,
          reason_code: reason_code,
          reason: REASONS.fetch(reason_code)
        )
      end

      def self.skipped(instruction, reason_code: :already_processed)
        new(
          instruction: instruction,
          status: :skipped,
          reason_code: reason_code,
          reason: REASONS.fetch(reason_code)
        )
      end

      def initialize(instruction:, status:, reason_code: nil, reason: nil)
        @instruction = instruction
        @status = status
        @reason_code = reason_code
        @reason = reason
      end

      def success?
        @status == :succeeded
      end

      def failure?
        @status == :failed
      end

      def skipped?
        @status == :skipped
      end
    end
  end
end
